alb-listeners.tf

```diff
@@ -1,39 +1,68 @@
-# If a certificate is provided, will redirect 80 to 443 and 443 will forward to target group
-resource "aws_lb_listener" "redirect_to_https" {
-  count             = var.certificate_arn != "" ? 1 : 0
+resource "aws_lb_listener" "http_redirects" {
+  for_each = {
+    for r in var.listeners : r.port => r
+    if r.action.type == "redirect" && r.protocol == "HTTP"
+  }
   load_balancer_arn = aws_lb.alb.arn
-  port              = "80"
-  protocol          = "HTTP"
+  port              = each.value.port
+  protocol          = each.value.protocol
   default_action {
     type = "redirect"
     redirect {
-      port        = "443"
-      protocol    = "HTTPS"
-      status_code = "HTTP_301"
+      host        = try(each.value.action.host, null)
+      path        = try(each.value.action.path, null)
+      port        = each.value.action.port
+      protocol    = each.value.action.protocol
+      query       = try(each.value.action.query, null)
+      status_code = try(each.value.action.status_code, "HTTP_301")
     }
   }
 }
-resource "aws_lb_listener" "forward_https" {
-  count             = var.certificate_arn != "" ? 1 : 0
+resource "aws_lb_listener" "https_redirects" {
+  for_each = {
+    for r in var.listeners : r.port => r
```

Comment:
I'd spell out `redirect_listener` instead of just `r`.  Or maybe add a comment "Filter out redirect to https listeners"